### PR TITLE
fix adapt: implement scale keyword for diameter/charge and 2d compatibility

### LINKED ISSUE #1508: [Feature Request] implement atom-style variables for fix adapt diameter
Dear developers,

Would it be possible to enable the use atom-style variables for the 'diameter' attribute of the 'fix adapt' command ? The same way 'fix setforce' accepts atom-style variable for example ? The present 'fix adapt' command only supports equal-style variables: all particles have the same diameter and it is not possible to adapt diameter for polydisperse granular systems.

This would be a very helpful feature as the only way to modify the diameter of polydisperse particles is currently to use 'set diameter' periodically in a loop or in a 'run xx every yy' which can become hefty when a smooth evolution of diameter is needed.

Thank you very much for your consideration.

### PR DESCRIPTION
**Pull request by jibril-b-coulibaly**
**Summary**

This PR implements 2 main changes to the `fix adapt` command:
- the `scale` keyword can apply to the parameters `diameter` and `charge`
- a `diameter/disc` parameter is defined so that mass changes associated with diameter changes can be consistent for 2d simulations

Bugs fixes and corresponding edits to the documentation are also performed.

**Related Issues**

This closes #1508 , enhancement I created to allow `fix adapt` to change the diameter of polydisperse granular assemblies. This is realized by implementing the `scale` capability for the `diameter` parameter. It is not as general as implementing atom-style variables as I initially suggested in issue #1508 but was much easier implementation and fits many polydisperse applications.

**Author(s)**

Jibril B. Coulibaly, Northwestern University, jibril.coulibaly@northwestern.edu

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

yes

**Implementation Notes**

To implement the `scale` keyword for `diameter` and `charge` parameters, the `FixAdapt::post_constructor()` function is modified so that the internal `Fix Store` are created even when the `reset no` option is used. This allows storing the initial diameter/charge and scaling them in the `FixAdapt::change_settings()` function, the same way `pair` and `bond` parameters are presently scaled.

The implementation of the `diameter/disc` parameter uses a `discflag` to determine the way particles mass should be modified and uses the disc area to define the mass in the `FixAdapt::change_settings()` and `FixAdapt::restore_settings()` functions.

Some non-compliant comments are currently left in the source files to facilitate the review of this PR.

[Documented input files](https://github.com/lammps/lammps/files/4331328/MWE_fix_adapt_PR.zip) showing the effects of the modifications proposed in this PR are enclosed:
 - `in.bug_post_constructor_reset_radius_charge_flag`: shows current bug in `FixAdapt::post_constructor()`
- `in.scale_diameter_charge`: demonstrates implementation of the `scale` keyword for `diameter` and `charge` parameters
- `in.diameter_disc`: demonstrates implementation of the 2d compatibility using `diameter/disc` parameter

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

### CODE REVIEW

**Comment by akohlmey** [comment]
thanks for your contribution. this apparently will need some careful review and i already noticed that there are some issues with included changes and some discussion points. I am writing this comment to acknowledge your submission but also to notify you that it will take some time until we will look into it, as other pending pull requests currently have a higher priority.

File: src/fix_adapt.cpp [hunk 1]
```diff
@@ -568,46 +573,49 @@ void FixAdapt::change_settings()
...
+						if (scaleflag) radius[i] = value * vec[i];
+						else radius[i] = 0.5*value;
+						if(discflag) rmass[i] = MY_PI * radius[i]*radius[i] * density;
+            else rmass[i] = 4.0*MY_PI/3.0 *
+                            radius[i]*radius[i]*radius[i] * density;
+          }
+
       } else if (ad->aparam == CHARGE) {
-        double *q = atom->q;
+        double *vec = fix_chg->vstore; // Get initial charge to use `scale` keyword
```
Comment (sjplimp) [inline src/fix_adapt.cpp:600]: don't use a trailing comment - put it on its own line

File: src/fix_adapt.cpp [hunk 2]
```diff
@@ -568,46 +573,49 @@ void FixAdapt::change_settings()
...
-              density = rmass[i] / (4.0*MY_PI/3.0 *
-                                    radius[i]*radius[i]*radius[i]);
-              radius[i] = 0.5*value;
-              rmass[i] = 4.0*MY_PI/3.0 *
-                radius[i]*radius[i]*radius[i] * density;
-            }
-        }
+        for (i = 0; i < nall; i++)
+          if (mask[i] & groupbit) {
+						if(discflag) density = rmass[i] / (MY_PI * radius[i]*radius[i]);
```
Comment (sjplimp) [inline src/fix_adapt.cpp:589]: indentation is wrong?  see comment below

File: src/fix_adapt.cpp [hunk 3]
```diff
@@ -568,46 +573,49 @@ void FixAdapt::change_settings()
       // also scale rmass to new value
 
       if (ad->aparam == DIAMETER) {
-        int mflag = 0;
-        if (atom->rmass_flag) mflag = 1;
+				/* `mflag` unnecessary ? the test `if(!atom->radius_flag)` in `FixAdapt::init()` should perevent `atom->rmass_flag == false`. Unless there can be combinations of atom styles with `radius` but without `rmass`
```
Comment (sjplimp) [inline src/fix_adapt.cpp:576]: ok to delete mflag here (and remove these comments) - but please
add a test for rmass flag (similar to radius_flag_ in init()

**Comment by akohlmey** [comment]
@jibril-b-coulibaly updated this branch to current master to resolve the merge conflict in the documentation of fix adapt. Please review and confirm if everything is still correct and we can proceed with merging this PR.

**Comment by jibril-b-coulibaly** [comment]
Thank you for the careful review @sjplimp .
@akohlmey Thank you for the conflict resolution. I reviewed the changes and confirm that everything is correct.

**Pull request merged**